internal/scheduler/preempting_queue_scheduler.go

```diff
@@ -0,0 +1,843 @@
+	return &Evictor{
+		jobRepo:         jobRepo,
+		priorityClasses: priorityClasses,
+		nodeFilter:      nodeFilter,
+		jobFilter: func(ctx context.Context, job interfaces.LegacySchedulerJob) bool {
+			if job.GetAnnotations() == nil {
+				log := ctxlogrus.Extract(ctx)
+				log.Warnf("can't evict job %s: annotations not initialised", job.GetId())
+				return false
+			}
+			priorityClassName := job.GetRequirements(priorityClasses).PriorityClassName
+			priorityClass, ok := priorityClasses[priorityClassName]
+			if !ok {
+				priorityClass = priorityClasses[defaultPriorityClass]
+			}
+			if priorityClass.Preemptible {
+				return true
+			}
+			return false
+		},
+		postEvictFunc: defaultPostEvictFunc,
+	}
+}
+
+// NewFilteredEvictor returns a new evictor that evicts all jobs for which jobIdsToEvict[jobId] is true
+// on nodes for which nodeIdsToEvict[nodeId] is true.
+func NewFilteredEvictor(
+	jobRepo JobRepository,
+	priorityClasses map[string]configuration.PriorityClass,
+	nodeIdsToEvict map[string]bool,
+	jobIdsToEvict map[string]bool,
+) *Evictor {
+	if len(nodeIdsToEvict) == 0 || len(jobIdsToEvict) == 0 {
+		return nil
+	}
+	return &Evictor{
+		jobRepo:         jobRepo,
+		priorityClasses: priorityClasses,
+		nodeFilter: func(_ context.Context, node *schedulerobjects.Node) bool {
+			shouldEvict := nodeIdsToEvict[node.Id]
+			return shouldEvict
+		},
+		jobFilter: func(_ context.Context, job interfaces.LegacySchedulerJob) bool {
+			shouldEvict := jobIdsToEvict[job.GetId()]
+			return shouldEvict
+		},
+		postEvictFunc: defaultPostEvictFunc,
+	}
+}
+
+// NewOversubscribedEvictor returns a new evictor that
+// for each node evicts all preemptible jobs of a priority class for which at least one job could not be scheduled
+// with probability perNodeEvictionProbability.
+func NewOversubscribedEvictor(
+	jobRepo JobRepository,
+	priorityClasses map[string]configuration.PriorityClass,
+	perNodeEvictionProbability float64,
+) *Evictor {
+	if perNodeEvictionProbability <= 0 {
+		return nil
+	}
+	// Populating overSubscribedPriorities relies on
+	// - nodeFilter being called once before all calls to jobFilter and
+	// - jobFilter being called for all jobs on that node before moving on to another node.
+	var overSubscribedPriorities map[int32]bool
+	prioritiesByName := configuration.PriorityByPriorityClassName(priorityClasses)
+	return &Evictor{
+		jobRepo:         jobRepo,
+		priorityClasses: priorityClasses,
+		nodeFilter: func(_ context.Context, node *schedulerobjects.Node) bool {
+			overSubscribedPriorities = make(map[int32]bool)
+			for p, rl := range node.AllocatableByPriorityAndResource {
+				if p < 0 {
+					// Negative priorities correspond to already evicted jobs.
+					continue
+				}
+				for _, q := range rl.Resources {
+					if q.Cmp(resource.Quantity{}) == -1 {
+						overSubscribedPriorities[p] = true
+						break
+					}
+				}
+			}
+			return len(overSubscribedPriorities) > 0 && rand.Float64() < perNodeEvictionProbability
```

Comment:
lets get a rand we can control in here so that we can test this

Reply:
Done

Reply:
Changed to
```
func NewOversubscribedEvictor(
	jobRepo JobRepository,
	priorityClasses map[string]configuration.PriorityClass,
	perNodeEvictionProbability float64,
	random *rand.Rand,
) *Evictor {
	if perNodeEvictionProbability <= 0 {
		return nil
	}
	if random == nil {
		random = rand.New(rand.NewSource(int64(time.Now().Nanosecond())))
	}
```